### PR TITLE
blog: add new post about current UIWebView updates

### PR DESCRIPTION
**Pull request by NiklasMerz**
### Platforms affected
iOS

### Motivation and Context

Recently a lot of people asked about UIWebView rejections and CORS issues. It might be a good idea to publish an update like we did for the build flag: https://cordova.apache.org/howto/2020/03/18/wkwebviewonly.html

### Description

I took the old UIWebView post and updated information from the [ios@6 releasenotes](https://cordova.apache.org/announcements/2020/06/01/cordova-ios-release-6.0.0.html).

**TODO**
* information about splashscreen change? (wait for release)
* more about CORS?

Looking for feedback. Just change or use PR suggestions ;-)

### CODE REVIEW

**Comment by breautek** [comment]
I think talking about cors is a good idea, I think it's the main one I see on slack nowadays, and many opened issues tend to be issues actually related to cors. I wrote about this specifically at https://breautek.com/2020/07/14/enabling-cors/

Feel free to pick apart anything that you feel make sense for Cordova.

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs.
+
+Some things have changed and new versions of cordova-ios have been released since the last post.
+
+<!--more-->
+
+## Update cordova-ios to version 6.0.0 or newer
+
+Please update to the latest cordova-ios version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into cordova-iOS and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
```
Comment (breautek) [inline www/_posts/2020-07-18-uiwebview-warning.md:23]:
```suggestion
Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
```

Comment (breautek) [reply www/_posts/2020-07-18-uiwebview-warning.md:23]: I think we normally emphasis package names using back-ticks? Ignore if I'm wrong

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+It is important to know that with the introduction of `WKURLSchemeHandler`, iOS 10 support has been dropped.
+
+## Still getting the warning?
+
+If you are still getting the warning, it is most likely one or more plugins in your project are still references `UIWebView`. You will need to identify which plugins and contact those plugin's developers through their support channel (plugin's repo). They will need to fix their plugins by either remove the references or wrap them with the new flag.
+
+## Using other WKWebView plugin
+
+The official Apache WKWebView plugin is no longer need with these cordova-ios versions since WKWebView has been integrated and UIWebView is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
```
Comment (breautek) [inline www/_posts/2020-07-18-uiwebview-warning.md:40]:
```suggestion
The official Apache WKWebView plugin is no longer need with these `cordova-ios` versions since WKWebView has been integrated and UIWebView is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
```

Comment (breautek) [reply www/_posts/2020-07-18-uiwebview-warning.md:40]: I think we normally emphasis package names using back-ticks? Ignore if I'm wrong

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+Some things have changed and new versions of cordova-ios have been released since the last post.
+
+<!--more-->
+
+## Update cordova-ios to version 6.0.0 or newer
+
+Please update to the latest cordova-ios version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into cordova-iOS and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
+
+Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Using a custom scheme to serve your app content through fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.
```
Comment (timbru31) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]: > Using a custom scheme to serve your app content through fixes CORS issues [...]

Is this valid English @breautek? I'd rephrase it to something like this:

> Serving your app content through a custom scheme fixes CORS issues [...]

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+```xml
+<preference name="scheme" value="app" />
+<preference name="hostname" value="localhost" />
+```
+
+It is important to know that with the introduction of `WKURLSchemeHandler`, iOS 10 support has been dropped.
+
+## Still getting the warning?
+
+If you are still getting the warning, it is most likely one or more plugins in your project are still references `UIWebView`. You will need to identify which plugins and contact those plugin's developers through their support channel (plugin's repo). They will need to fix their plugins by either remove the references or wrap them with the new flag.
```
Comment (timbru31) [inline www/_posts/2020-07-18-uiwebview-warning.md:36]:
```suggestion
If you are still getting the warning, it is most likely one or more plugins in your project still references `UIWebView`. You will need to identify which plugins contain `UIWebView` code and contact those plugin's developers through their support channel (for instance their plugin's repo). They will need to fix their plugins by either removing the references or wrap them with the new flag.
```

File: www/_posts/2020-07-18-uiwebview-warning.md (hunk 2 above)
Comment (timbru31) [inline www/_posts/2020-07-18-uiwebview-warning.md:40]:
```suggestion
The official Apache WKWebView plugin is no longer needed with these `cordova-ios` versions since `WKWebView` has been integrated and `UIWebView` is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
```

File: www/_posts/2020-07-18-uiwebview-warning.md (hunk 3 above)
Comment (breautek) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]: We may also want to note that changing schemes will cause clients to lose access to their current web storage, such as local storage, indexed DB, etc.

It's possible to migrate this data but Cordova doesn't currently handle this.

Comment (NiklasMerz) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]:
```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.
```

Comment (NiklasMerz) [reply www/_posts/2020-07-18-uiwebview-warning.md]: Done

**Comment by NiklasMerz** [comment]
> I think talking about cors is a good idea, I think it's the main one I see on slack nowadays, and many opened issues tend to be issues actually related to cors. I wrote about this specifically at https://breautek.com/2020/07/14/enabling-cors/
> 
> Feel free to pick apart anything that you feel make sense for Cordova.

That's a great post. I would just put a link to it at the end of this Cordova post. What do you think?

**Comment by breautek** [comment]
> information about splashscreen change?

I think the latest release of the splashscreen plugin includes the engine check to avoid installing the plugin on cordova-ios >= 6. However, @erisu was experiencing some clobber issues with the javascript... still might be worth mentioning that upgrading to cordova-ios@6 will require an upgrade to the `cordova-plugin-splashscreen` plugin.

File: www/_posts/2020-07-18-uiwebview-warning.md (hunk 3 above)
Comment (NiklasMerz) [reply www/_posts/2020-07-18-uiwebview-warning.md]: Yeah a good idea to warn about that. Just pushed one sentence.

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+<!--more-->
+
+## Update cordova-ios to version 6.0.0 or newer
+
+Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
+
+Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin your web code runs and you will loose access to web storage, such as local storage, indexed DB, etc.
```
Comment (breautek) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]:
```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.
```

**Comment by NiklasMerz** [comment]
> > information about splashscreen change?
> 
> I think the latest release of the splashscreen plugin includes the engine check to avoid installing the plugin on cordova-ios >= 6. However, @erisu was experiencing some clobber issues with the javascript... still might be worth mentioning that upgrading to cordova-ios@6 will require an upgrade to the `cordova-plugin-splashscreen` plugin.

A new version for the splashscreen plugin is currently being voted on. I think it fixes all issues with the latest cordova-ios versions. We should publish this post once that is released.

File: www/_posts/2020-07-18-uiwebview-warning.md (hunk 5 above)
Comment (breautek) [reply www/_posts/2020-07-18-uiwebview-warning.md:25]: typo and grammar improvements.

**Comment by NiklasMerz** [comment]
Thanks for the great suggestions folks.

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+If you are still getting the warning, it is most likely one or more plugins in your project still references `UIWebView`. You will need to identify which plugins contain `UIWebView` code and contact those plugin's developers through their support channel (for instance their plugin's repo). They will need to fix their plugins by either removing the references or wrap them with the new flag.
+
+## Using other WKWebView plugin
+
+The official Apache WKWebView plugin is no longer needed with these `cordova-ios` versions since `WKWebView` has been integrated and `UIWebView` is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
+
+## CORS issues
+
+If you experience any CORS issues Cordova PMC member Norman Breau published a [post on his](https://breautek.com/2020/07/14/enabling-cors/) with a good explanation of CORS.
```
Comment (timbru31) [inline www/_posts/2020-07-18-uiwebview-warning.md:44]:
```suggestion
If you experience any CORS issues Cordova PMC member Norman Breau published a [post on his blog](https://breautek.com/2020/07/14/enabling-cors/) with a good explanation of CORS.
```

**Comment by breautek** [comment]
Thank you for preparing this PR @NiklasMerz !

I think all we need now is the splash screen release.

PS. I think I need to sort out my email client because I think Cordova emails are going to my spam/junk folder again, even though it's configured not to...

**Comment by NiklasMerz** [comment]
> PS. I think I need to sort out my email client because I think Cordova emails are going to my spam/junk folder again, even though it's configured not to...

This is strange for me,too. I cant find the splashscreen thread in my inbox, too. But other mails sent to the list are there

File: www/_posts/2020-07-18-uiwebview-warning.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+<!--more-->
+
+## Update cordova-ios to version 6.0.0 or newer
+
+Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
+
+Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.
```
Comment (timbru31) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]: Is bold formatting supported? I'd argue that the sentence
> Bear in mind [...]
should be bold to catch the reader's attention because it's well hidden in the text.

Comment (NiklasMerz) [inline www/_posts/2020-07-18-uiwebview-warning.md:25]:
```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. **Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.**
```

Good point. Bold should work.

**Comment by NiklasMerz** [comment]
Splashscreen is released. Let's publish.

**Comment by jcesarmobile** [comment]
It would be good to mention the `convertFilePath` helper method since a few people has created issues or asked on github about not being able to load file urls when using a scheme

I created an issue to document it, but probably makes sense to mention it here  too https://github.com/apache/cordova-ios/issues/947
